stdlib/sql/hdb.go

```diff
@@ -0,0 +1,211 @@
+		default:
+			execute.PanicUnknownType(flux.TInvalid)
+		}
+	}
+	return row, nil
+}
+
+func (m *HdbRowReader) InitColumnNames(names []string) {
+	m.columnNames = names
+}
+
+func (m *HdbRowReader) InitColumnTypes(types []*sql.ColumnType) {
+	fluxTypes := make([]flux.ColType, len(types))
+	for i := 0; i < len(types); i++ {
+		switch types[i].DatabaseTypeName() {
+		case "TINYINT", "SMALLINT", "INTEGER", "BIGINT":
+			fluxTypes[i] = flux.TInt
+		case "REAL", "DOUBLE", "DECIMAL":
+			fluxTypes[i] = flux.TFloat
+		case "TIMESTAMP": // not exactly correct (see Notes)
+			fluxTypes[i] = flux.TTime
+		default:
+			fluxTypes[i] = flux.TString
+		}
+	}
+	m.columnTypes = fluxTypes
+	m.sqlTypes = types
+}
+
+func (m *HdbRowReader) ColumnNames() []string {
+	return m.columnNames
+}
+
+func (m *HdbRowReader) ColumnTypes() []flux.ColType {
+	return m.columnTypes
+}
+
+func (m *HdbRowReader) SetColumnTypes(types []flux.ColType) {
+	m.columnTypes = types
+}
+
+func (m *HdbRowReader) SetColumns(i []interface{}) {
+	m.columns = i
+}
+
+func (m *HdbRowReader) Close() error {
+	if m.CloseFunc != nil {
+		return m.CloseFunc()
+	}
+	if err := m.Cursor.Err(); err != nil {
+		return err
+	}
+	return m.Cursor.Close()
+}
+
+func NewHdbRowReader(r *sql.Rows) (execute.RowReader, error) {
+	reader := &HdbRowReader{
+		Cursor: r,
+	}
+	cols, err := r.Columns()
+	if err != nil {
+		return nil, err
+	}
+	reader.InitColumnNames(cols)
+
+	types, err := r.ColumnTypes()
+	if err != nil {
+		return nil, err
+	}
+	reader.InitColumnTypes(types)
+
+	return reader, nil
+}
+
+var fluxToHdb = map[flux.ColType]string{
+	flux.TFloat:  "DOUBLE",
+	flux.TInt:    "BIGINT",
+	flux.TString: "NVARCHAR(5000)", // 5000 is the max
+	flux.TBool:   "BOOLEAN",
+	flux.TTime:   "TIMESTAMP", // not exactly correct (see Notes)
+}
+
+// HdbTranslateColumn translates flux colTypes into their corresponding SAP HANA column type
+func HdbColumnTranslateFunc() translationFunc {
+	return func(f flux.ColType, colName string) (string, error) {
+		s, found := fluxToHdb[f]
+		if !found {
+			return "", errors.Newf(codes.Internal, "SAP HANA does not support column type %s", f.String())
+		}
+		return colName + " " + s, nil
+	}
+}
+
+// Template for conditional query by table existence check
+var hdbDoIfTableNotExistsTemplate = `DO
+BEGIN
+    DECLARE X_EXISTS INT = 0;
+    SELECT COUNT(*) INTO X_EXISTS FROM TABLES %s;
+    IF :X_EXISTS = 0
+    THEN
+        %s;
+    END IF;
+END;
+`
+
+// Adds SAP HANA specific table existence check to CREATE TABLE statement
+func hdbAddIfNotExist(table string, query string) string {
+	var where string
+	parts := strings.SplitN(table, ".", 2)
+	if len(parts) == 2 { // fully-qualified table name
+		where = fmt.Sprintf("WHERE SCHEMA_NAME=UPPER('%s') AND TABLE_NAME=UPPER('%s')", parts[0], parts[1])
+	} else { // table in user default schema
+		where = fmt.Sprintf("WHERE TABLE_NAME=UPPER('%s')", table)
+	}
+	return fmt.Sprintf(hdbDoIfTableNotExistsTemplate, where, query)
```

Comment:
I'm a little worried that someone could craft a table name that contains characters like `'; ...` or something and do something nefarious. [example](https://xkcd.com/327/)

Do you have any thoughts on this? For some of the places where we use `%s` I imagine that we could use parameters `?` instead?

Reply:
I could not get ? placeholder to work with the SQL DO script (`SQL Error 1287 - identifier must be declared`), but I changed it to use declared variables for schema and table. The resulting query now looks like eg.
```
DO
BEGIN
    DECLARE SCHEMA_NAME NVARCHAR(11) = 'bike_stores';
    DECLARE TABLE_NAME NVARCHAR(5) = 'xcopy';
    DECLARE X_EXISTS INT = 0;
    SELECT COUNT(*) INTO X_EXISTS FROM TABLES WHERE SCHEMA_NAME=UPPER(:SCHEMA_NAME) AND TABLE_NAME=UPPER(:TABLE_NAME);
    IF :X_EXISTS = 0
    THEN
        CREATE TABLE bike_stores.xcopy (NOTE NVARCHAR(5000),...);
    END IF;
END;
```

Reply:
I think this is better but it still seems like the `CREATE TABLE` statement allows for an injection.

If I were to do something like `sql.to(table: "foo (i INTEGER); DROP TABLE bar; ...")`, then there's still a problem, if I'm understanding right.

What if you always quoted the identifiers in the CREATE TABLE statement:
```
CREATE TABLE "bike_stores"."xcopy" ("NOTE" NVARCHAR(5000),...);
```
And if any of the identifiers contain a `"`, you can either escape it, or produce a user error. Does that make sense?

Reply:
Improved, I hope :) Since by default object names are converted to uppercase by HDB unless quoted, and quotes are used to escape possibly malign SQL code here and not to mark the value as case-sensitive, uppercase is assumed and used for **new** output table name, which I think is conforming and safe.

Table created like bellow then allows a user to insert data both with any of the following statements:

`INSERT INTO bike_stores.orders_copy_q2 ... (note, ...) VALUES (...)`
`INSERT INTO BIKE_STORES.ORDERS_COPY_Q2 ... (NOTE, ...) VALUES (...)`.
`INSERT INTO "BIKE_STORES"."ORDERS_COPY_Q2" ... ("NOTE", ...) VALUES (...)`.

The if-not-exist-then-create now looks like this:

```DO
BEGIN
    DECLARE SCHEMA_NAME NVARCHAR(11) = 'BIKE_STORES';
    DECLARE TABLE_NAME NVARCHAR(14) = 'ORDERS_COPY_Q2';
    DECLARE X_EXISTS INT = 0;
    SELECT COUNT(*) INTO X_EXISTS FROM TABLES WHERE SCHEMA_NAME=ESCAPE_DOUBLE_QUOTES(:SCHEMA_NAME) AND TABLE_NAME=ESCAPE_DOUBLE_QUOTES(:TABLE_NAME);
    IF :X_EXISTS = 0
    THEN
        CREATE TABLE "BIKE_STORES"."ORDERS_COPY_Q2" ("NOTE" NVARCHAR(5000),...);
    END IF;
END;

Reply:
This is much better. Thanks for the attention to this.